File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.py
## time:o(n)
## space: o(n)
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        
        ## count for counting subarrays in the input array
        count = 0
        
        sum = 0
        
        ## create a dic to store the frequency as values
        dic = {}
        
        ## initialize 1st value as {0:1}
        dic[0] = 1
        
        for i in range(len(nums)):
            
            sum = sum + nums[i]
            
            ## check sum-k already exists in dic or not  , if exists then increament count
            if(sum-k) in dic:
                count = count + dic[sum-k]
                
            ## update dic
            dic[sum] = dic.get(sum,0) + 1
            
        return count    
        
        
        
        

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        
        int count =0;
        int sum = 0;
        
        unordered_map<int , int> dic;
        
        dic[0] = 1;
        
        for(int i=0;i<nums.size();i++){
            sum = sum+nums[i];
            
            if(dic.find(sum - k) != dic.end()){
                count = count + dic[sum-k];
            }
            dic[sum]++;
        }
        return count;
    }
};

/*
int main() {
    Solution solution;
    
    // Input array and k value
    vector<int> nums = {1, 2, 3};
    int k = 3;
    
    // Call the subarraySum function and print the result
    int result = solution.subarraySum(nums, k);
    cout << result << endl;
    
    return 0;
}
*/